### DISCUSSION TITLE
Test report after the latest AI bonus changes (1.1.0 8814)

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#8840] [General]
AI bonus in the test:

> @oleg-derevenetz,
> 
> > > Hi @LeHerosInconnu
> > 
> > > Castle type A, bonus 10 000 gold and X rare resources.
> > > Castle type B, bonus 5 000 gold no resources.
> > 
> > Currently per-castle resource bonuses (on Expert and Impossible difficulties) are as follows:
> > 
> > * +1 Crystal per day for each Barbarian castle;
> > * +1 Mercury per day for each Sorceress castle;
> > * +1 Sulfur & +1000 Gold per day for each Warlock castle;
> > * +1 Gem & +1000 Gold per day for each Wizard castle;
> > 
> > The above bonuses are applied only if all the following conditions are met:
> > 
> > * The corresponding castle has a level 6 dwelling;
> > * The corresponding kingdom has a Marketplace in at least one of its castles or has at least one captured mine of the appropriate type.
> > 
> > In addition, 1000 gold is always awarded for each castle in the kingdom regardless of its type and dwellings (separately, in addition to "usual" (non-AI-specific) gold income from castle).
> > 
> Okay, thanks.

_Originally posted by @LeHerosInconnu in https://github.com/ihhub/fheroes2/issues/8402#issuecomment-2156099809_

Test on Dominion scenario (PoL), XL sized map, impossible difficulty.

AI players are less aggressive than in my previous tests (this is normal, as they have fewer bonuses).
I was able to execute my plan without too much difficulty (conquer the castles of the orange player (wizard), then those of the yellow player (warlock) and complete the scenario more or less quickly, depending on whether the dimension door spell is available).
The AI players' cities are less developed (again, this is normal, they have fewer bonuses).
I haven't completed the scenario, it's not necessary, the AI won't be able to take over at this stage of the game.
I have a powerful hero with titans and black dragons at my disposal, so it's just a matter of time before I finish the scenario.
Conquered castles have no (or very few) unrecruited creatures, so that's good.

<details>
<summary>Creatures available after conquest: (click to expand)</summary>
<br>

Orange AI player's first city captured (castle), month 2, week 2, day 5 (After orange 001.sav).
![After orange 001](https://github.com/ihhub/fheroes2/assets/43583994/28338eb7-78b2-42bc-9d6e-ed297096c5ba)

Orange AI player's second city captured (castle), month 2, week 2, day 5 (After orange 002.sav).
![After orange 002](https://github.com/ihhub/fheroes2/assets/43583994/3135ecb3-c43d-4660-b5be-ad907b945c93)

Orange AI player's third city captured (town), month 2, week 4, day 5 (After orange 003.sav).
No creatures available.
![After orange 003](https://github.com/ihhub/fheroes2/assets/43583994/e363dc0d-92c4-4493-8f39-e8b51f1afe6c)

Orange AI player's fourth city captured (castle), month 2, week 4, day 6 (After orange 004.sav).
![After orange 004](https://github.com/ihhub/fheroes2/assets/43583994/dc89cd9d-8d58-4dde-8fce-295805be1159)

Green AI player's first city captured (castle), month 3, week 1, day 1 (After green 001.sav).
It's okay. It's day 1 of the week.
![After green 001](https://github.com/ihhub/fheroes2/assets/43583994/b1053575-76d1-40d9-a66f-54e818b58dd9)

Yellow AI player's first city captured (castle), month 3, week 3, day 5 (After yellow 001.sav).
![After yellow 001](https://github.com/ihhub/fheroes2/assets/43583994/9fcaa302-8033-44ab-8096-f06aa2c0eb58)

Yellow AI player's second city captured (castle), month 3, week 3, day 5 (After yellow 002.sav).
![After yellow 002](https://github.com/ihhub/fheroes2/assets/43583994/426cfe5b-db92-4ac1-b4aa-5b7dc99dd24b)

Yellow AI player's third city captured (castle), month 3, week 3, day 6 (After yellow 003.sav).
![After yellow 003](https://github.com/ihhub/fheroes2/assets/43583994/0a552fd8-21e3-41ab-ac7e-847d77430b72)

Red AI player's first city captured (castle), month 3, week 3, day 1 (After red 001.sav).
It's okay. It's day 1 of the week.
![After red 001](https://github.com/ihhub/fheroes2/assets/43583994/fcb250fe-f240-45a6-8ba2-cc4dca59f4e3)

Red AI player's second city captured (castle), month 3, week 3, day 6 (After red 002.sav).
It's okay. The castle was defended by other creatures and my hero was relatively far away before the attack.
![After red 002](https://github.com/ihhub/fheroes2/assets/43583994/7210ed49-604c-454d-b389-38a3512443ac)

Red AI player's first city recaptured (castle), month 3, week 3, day 6 (After red recap 001.sav).
![After red recap 001](https://github.com/ihhub/fheroes2/assets/43583994/47d5c4f6-1545-4991-998c-9cedb3080704)

Red AI player's third city captured (castle), month 3, week 3, day 7 (After red 003.sav).
![After red 003](https://github.com/ihhub/fheroes2/assets/43583994/3ac135d3-5b81-4a93-8459-34d565046e60)

Red AI player's second city recaptured (castle), month 3, week 4, day 2 (After red recap 002.sav).
![After red recap 002](https://github.com/ihhub/fheroes2/assets/43583994/3bd5103c-f703-46c0-afc2-f7ed4765be28)

</details>

As for the bonus according to castle type, I think it would be preferable to give a "fair" bonus for each type of castle, in order to preserve the initial "balance" of the game.
At the moment, the bonus is given according to the type of level six creature.
This is not very fair, as castles with powerful creatures get a higher bonus.
So it's a "double bonus", firstly they have the most powerful creatures, and secondly they receive a higher bonus to recruit them.

In the game, each type of castle initially receives the same bonus, 1000 gold for the castle and 250 gold for the statue.
The extra 500 gold for the warlock is given by its special building, but there are also two upgrades to build for the dwelling of level six creature.

I think it would be appropriate to adjust the bonus to the "same level" for all castle types (based on the highest bonus value).
This way:
- +1 unit of crystal and +1000 gold per day for each knight castle
- +1 unit of crystal and +1000 gold per day for each barbarian castle
- +1 unit of mercury and +1000 gold per day for each sorceress castle
- +1 unit of sulfur and +1000 gold per day for each necromancer castle
- +1 unit of gems and +1000 gold per day for each wizard castle
- +1 unit of sulfur and +1000 gold per day for each warlock castle

Everything else remains unchanged.
For weaker castles, the adjustment would make it possible to compensate by buying other heroes, recruiting creatures from the adventure map's dwellings, etc.

I think we can wait for user feedback (and testing) before making any further changes (apart from the "fair" adjustment).

Here are the save files:
[After conquest 01.zip](https://github.com/user-attachments/files/15858064/After.conquest.01.zip)

**Comment by oleg-derevenetz** [comment]
> I think it would be appropriate to adjust the bonus to the "same level" for all castle types (based on the highest bonus value).
> This way:
> 
> +1 unit of crystal and +1000 gold per day for each knight castle
> +1 unit of crystal and +1000 gold per day for each barbarian castle
> +1 unit of mercury and +1000 gold per day for each sorceress castle
> +1 unit of sulfur and +1000 gold per day for each sorceress castle
> +1 unit of gems and +1000 gold per day for each wizard castle
> +1 unit of sulfur and +1000 gold per day for each warlock castle

The point of the changes in terms of additional money was so that AI could hire the entire weekly unit growth (at least on Expert and Impossible difficulties), and now it can do it. Some castles have more expensive units than other castles, so some types of castles get bigger bonuses than other types of castles. If we give all castles the same bonuses, then AI players with castles of certain types will have a shortage of funds to buy units, while AI players with castles of other types will have an excess of funds (as far as I remember, @idshibanov was concerned that there would be no undisguised distortions in the development of different AI players, depending on the types of their castles). In addition, it makes no sense to give additional funds for castles in which units do not require rare resources for hiring (Knight & Necromancer).

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz,

> > I think it would be appropriate to adjust the bonus to the "same level" for all castle types (based on the highest bonus value).
> > This way:
> > +1 unit of crystal and +1000 gold per day for each knight castle
> > +1 unit of crystal and +1000 gold per day for each barbarian castle
> > +1 unit of mercury and +1000 gold per day for each sorceress castle
> > +1 unit of sulfur and +1000 gold per day for each necromancer castle
> > +1 unit of gems and +1000 gold per day for each wizard castle
> > +1 unit of sulfur and +1000 gold per day for each warlock castle
> 
> The point of the changes in terms of additional money was so that AI could hire the entire weekly unit growth (at least on Expert and Impossible difficulties), and now it can do it. Some castles have more expensive units than other castles, so some types of castles get bigger bonuses than other types of castles. If we give all castles the same bonuses, then AI players with castles of certain types will have a shortage of funds to buy units, while AI players with castles of other types will have an excess of funds (as far as I remember, @idshibanov was concerned that there would be no undisguised distortions in the development of different AI players, depending on the types of their castles). In addition, it makes no sense to give additional funds for castles in which units do not require rare resources for hiring (Knight & Necromancer).

The aim is to maintain the balance of the original game, also in relation to the different map sizes.

Four crusaders don't stand a chance against three titans.
For example, on a medium sized map, the knight will be penalized if the bonus given is not of equivalent value for all castle types.
Weaker castles need to be able to compensate for this by other means than purely recruiting their level six creatures.
That's why a fair bonus is necessary (one unit of the most used rare resource, and 1000 gold for all castle types).

Yes, creature recruitment has been improved, but that wouldn't mean altering the balance of the original game by giving different bonuses to different factions.
Originally, all castle types have the same income of 1250 gold for common buildings.
This equality should also be preserved in the case of bonuses.

For the necromancer's castle, sulfur is not directly necessary for the level six creature, but we also need to think about the development of other captured neutral castles of the same type to be able to have larger necromancer armies.
For the necromancer, sulfur is needed to build the mausoleum, a prerequisite for the construction of the bone dragon's dwelling.
It's all part of the game, and a way of compensating for weak castles that are developed more quickly.

**Reply by oleg-derevenetz** [reply]
> Yes, creature recruitment has been improved, but that wouldn't mean altering the balance of the original game by giving different bonuses to different factions.
> Originally, all castle types have the same income of 1250 gold for common buildings.
This equality should also be preserved in the case of bonuses.

As far as I remember, the task was originally stated as "not to allow AI to have unoccupied units in castles, which an enemy player can buy out after occupying the castle to his own advantage". Obviously, with such a task statement, castles with stronger units gain an advantage.

> For the necromancer's castle, sulfur is not directly necessary for the level six creature, but we also need to think about the development of other captured neutral castles of the same type to be able to have larger necromancer armies.
For the necromancer, sulfur is needed to build the mausoleum, a prerequisite for the construction of the bone dragon's dwelling.

I suspect that this is precisely the distortion of the speed of development of castles of a certain types that @idshibanov was concerned about.

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz,

> > Yes, creature recruitment has been improved, but that wouldn't mean altering the balance of the original game by giving different bonuses to different factions.
> > Originally, all castle types have the same income of 1250 gold for common buildings.
> > This equality should also be preserved in the case of bonuses.
> 
> As far as I remember, the task was originally stated as "not to allow AI to have unoccupied units in castles, which an enemy player can buy out after occupying the castle to his own advantage". Obviously, with such a task statement, castles with stronger units gain an advantage.

The original game is designed so that some castles are stronger than others, with different advantages and disadvantages depending on the size of the map.
The way in which bonuses are awarded breaks with the original gameplay.

> > For the necromancer's castle, sulfur is not directly necessary for the level six creature, but we also need to think about the development of other captured neutral castles of the same type to be able to have larger necromancer armies.
> > For the necromancer, sulfur is needed to build the mausoleum, a prerequisite for the construction of the bone dragon's dwelling.
> 
> I suspect that this is precisely the distortion of the speed of development of castles of a certain types that @idshibanov was concerned about.

It's part of the original game play that some castles are easier to develop than others.

**Reply by oleg-derevenetz** [reply]
I suggest to listen @idshibanov's opinion regarding this.

**Comment by zascode** [comment]
The castle towers shot at the attacking iron golems instead of at the ranged attacking units, which was standard in the original game and made a lot more sense. Would be nice if AI could be switched between the original version and the new one.

**Comment by LeHerosInconnu** [comment]
There is an option for this planned in my mod: https://github.com/ihhub/fheroes2/discussions/10102.
I close the discussion.
